File: services/edge-api/src/repository.ts
// Repository layer: raw parameterized SQL + explicit row → domain mapping, nothing
// else. The proving read consumes platform-owned data only — the goose migration
// ledger (003) and catalog functions — introducing zero product schema (research E2).
import { query } from "./lib/db";
import type { PlatformStatus } from "./types";

const qStatus = `
SELECT current_database()                                    AS database_name,
       now()                                                 AS database_time,
       COALESCE((SELECT MAX(version_id)
                   FROM goose_db_version
                  WHERE is_applied), 0)                      AS migration_version,
       (SELECT COUNT(*)
          FROM goose_db_version
         WHERE is_applied AND version_id > 0)                AS migrations_applied
`;

// Wire shape of qStatus — BIGINTs arrive as strings from the driver; mapped
// explicitly below and never exported.
interface StatusRow {
  database_name: string;
  database_time: Date;
  migration_version: string | number;
  migrations_applied: string | number;
}

export interface StatusRepository {
  status(): Promise<Omit<PlatformStatus, "environment">>;
}

export const statusRepository: StatusRepository = {
  async status() {
    const result = await query<StatusRow>(qStatus);
    const row = result.rows[0];
    if (!row) throw new Error("platformstatus: status query returned no row");

    return {
      databaseName: row.database_name,
      databaseTime: row.database_time,
      // Ledger version ids are 14-digit timestamps — safely inside Number range.
      migrationVersion: Number(row.migration_version),
      migrationsApplied: Number(row.migrations_applied),
    };
  },
};

// pingDatabase is the health check's dependency probe (2s budget enforced by the
// caller; SELECT 1 keeps it free of any table dependency).
export async function pingDatabase(): Promise<void> {
  await query("SELECT 1");
}

File: apis/core-api/internal/features/platformstatus/repository.go
// Repository layer: SQL only (ARCHITECTURE.md). The proving read consumes
// platform-owned data exclusively — the goose migration ledger (003's Migration
// Ledger) plus catalog functions — introducing zero product schema (research E2).
package platformstatus

import (
	"context"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"

	"github.com/effyshopping/effy/services/core-api/internal/platform/db"
)

// qStatus reads everything in one round trip. version 0 is goose's baseline marker
// row, excluded from the applied count.
const qStatus = `
SELECT current_database()                                    AS database_name,
       now()                                                 AS database_time,
       COALESCE((SELECT MAX(version_id)
                   FROM goose_db_version
                  WHERE is_applied), 0)                      AS migration_version,
       (SELECT COUNT(*)
          FROM goose_db_version
         WHERE is_applied AND version_id > 0)                AS migrations_applied
`

// statusRow is the wire shape of qStatus; it never leaves this file (mapped
// explicitly to the domain model below).
type statusRow struct {
	DatabaseName      string    `db:"database_name"`
	DatabaseTime      time.Time `db:"database_time"`
	MigrationVersion  int64     `db:"migration_version"`
	MigrationsApplied int64     `db:"migrations_applied"`
}

type Repository struct {
	db db.DBTX
}

func NewRepository(dbtx db.DBTX) *Repository {
	return &Repository{db: dbtx}
}

func (r *Repository) Status(ctx context.Context) (PlatformStatus, error) {
	rows, err := r.db.Query(ctx, qStatus)
	if err != nil {
		return PlatformStatus{}, fmt.Errorf("platformstatus: query: %w", err)
	}

	row, err := pgx.CollectExactlyOneRow(rows, pgx.RowToStructByName[statusRow])
	if err != nil {
		// Includes 42P01 when the goose ledger doesn't exist yet (003 db-up pending) —
		// the service maps this to an internal error; the cause lands in the log only.
		return PlatformStatus{}, fmt.Errorf("platformstatus: scan: %w", err)
	}

	return PlatformStatus{
		DatabaseName:      row.DatabaseName,
		DatabaseTime:      row.DatabaseTime.UTC(),
		MigrationVersion:  row.MigrationVersion,
		MigrationsApplied: row.MigrationsApplied,
	}, nil
}
